### ISSUE TITLE
[TF 2.0] Respect masking in Keras loss reduction (i.e., support an equivalent of default TF 1.0 loss reduction SUM_OVER_NONZERO_WEIGHTS)

### ISSUE THREAD

**Issue by foxik** [tensorflow/tensorflow#26784]
**System information**
- TensorFlow version (you are using): **TF 2.0.0a0**
- Are you willing to contribute it (Yes/No): **Yes**

**Describe the feature and the current behavior/state.**

In TF 1.0, default loss reduction was `SUM_OVER_NONZERO_WEIGHTS`. For NLP with sequences on inputs, it normalized losses by number of _valid elements_ (i.e., sum of non-padding words in input sentences).

With Keras API, default reduction `SUM_OVER_BATCH_SIZE` does not respect masking, so if a batch of sequences is passed on input, it is normalized by total batch size including padding (masked) elements. No reduction in Keras is available which would recover the previous TF 1.0 default `SUM_OVER_NONZERO_WEIGHTS` reduction.

**Will this change the current api? How?**

My proposal is to respect masking in the `SUM_OVER_BATCH_SIZE` reduction. If a mask is set, then `BATCH_SIZE` should probably correspond to the number of unmasked elements anyway.

Alternatively, a new reduction `SUM_OVER_MASKED_BATCH_SIZE` could be added.

**Who will benefit with this feature?**

I think anyone using masked losses, especially if _upgrading from TF 1_.

**Comment by Saduf2019** [comment]
@foxik Could you please confirm if you are still facing the issue.

**Comment by foxik** [comment]
Yes. TF 2.1.0 still evaluates
```python
mse = tf.keras.losses.MeanSquaredError(reduction=tf.keras.losses.Reduction.SUM_OVER_BATCH_SIZE)
mse([[0.], [0.], [0.]], [[1.], [1.], [1.]], sample_weight=[1, 0, 0])
```
to `0.333...` instead of to `1`.

At this point I am not proposing to change behaviour of `SUM_OVER_BATCH_SIZE`, which is already part of API guarantee -- but I propose to add
```python
tf.keras.losses.Reduction.SUM_OVER_NONZERO_WEIGHTS
```
which would compute the reduction by dividing by the number of non-zero weights, exactly what the default of TF 1 was.

**Comment by pavithrasv** [comment]
@foxik would you like to send us a PR with this change?

**Comment by foxik** [comment]
@pavithrasv Well, for a single-CPU/single-GPU training, the change is trivial. But for a non-trivial distribution strategy it will take some thinking how to implement it correctly. Also there seems to be several places in Keras which manually handle SUM_OVER_BATCH_SIZE, which will probably have to be adapted -- overall I do not feel confident I can put all the pieces together.

**Comment by foxik** [comment]
Closing the issue. With `train_step` one can override it and transform loss calls on higher-dimensional data (e.g., 2D loss in case of sequences as batch examples) to calls on one-dimensional data (e.g., concatenating valid words only).

Furthermore, with RaggedTensors, padding values might be handled directly, instead of passing masks as weihts.
